File: sum_root_leaf_numbers.java
/**
 * Time COmplexity: O(n) n-all nodes in the tree
 * Space Complexity - o(h) height of the tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result;
    public int sumNumbers(TreeNode root) {
        helper(root,0);
        return result;
    }
    public void helper(TreeNode node, int sum){
        if(node == null) return;

        sum = (sum*10) + node.val;
        if(node.left == null && node.right == null) result += sum;
        helper(node.left,sum);
        helper(node.right,sum);
    }
}

File: sumrootleafnumbers.py
##Time Complexity : O(n)
##Space Complexity : O(h), h-height of the tree
##Did this code successfully run on Leetcode : Yes
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(current, num):
            if not current:
                return 0
            num = num * 10 + current.val
            if not current.left and not current.right:
                return num
            return dfs(current.left, num) + dfs(current.right, num)
        return dfs(root, 0)